worlds/mm2/rules.py

```diff
@@ -0,0 +1,297 @@
+                        if boss == 9:
+                            if weapon != 3:
+                                # Atomic Fire and Crash Bomber cannot be Picopico-kun's only weakness
+                                world.weapon_damage[weapon][boss] = 0
+                                weakness = world.random.choice((2, 3, 4, 5, 7, 8))
+                                world.weapon_damage[weakness][boss] = minimum_weakness_requirement[weakness]
+                        elif boss == 11:
+                            if weapon == 1:
+                                # Atomic Fire cannot be Boobeam Trap's only weakness
+                                world.weapon_damage[weapon][boss] = 0
+                                weakness = world.random.choice((2, 3, 4, 5, 6, 7, 8))
+                                world.weapon_damage[weakness][boss] = minimum_weakness_requirement[weakness]
+                        else:
+                            world.weapon_damage[weapon][boss] = minimum_weakness_requirement[weapon]
+            starting = world.options.starting_robot_master.value
+            world.weapon_damage[0][starting] = 1
+
+        for p_boss in world.options.plando_weakness:
+            for p_weapon in world.options.plando_weakness[p_boss]:
+                if not any(w for w in world.weapon_damage
+                           if w != p_weapon and world.weapon_damage[w][bosses[p_boss]] > minimum_weakness_requirement[w]):
+                    # we need to replace this weakness
+                    weakness = world.random.choice([key for key in world.weapon_damage if key != p_weapon])
+                    world.weapon_damage[weakness][bosses[p_boss]] = minimum_weakness_requirement[weakness]
+                world.weapon_damage[weapons_to_id[p_weapon]][bosses[p_boss]] \
+                    = world.options.plando_weakness[p_boss][p_weapon]
+
+        if world.weapon_damage[0][world.options.starting_robot_master.value] < 1:
+            world.weapon_damage[0][world.options.starting_robot_master.value] = 1
+
+        # final special case
+        # There's a vanilla crash if Time Stopper kills Wily phase 1
+        # There's multiple fixes, but ensuring Wily cannot take Time Stopper damage is best
+        if world.weapon_damage[8][12] > 0:
+            world.weapon_damage[8][12] = 0
+
+        # weakness validation, it is better to confirm a completable seed than respect plando
+        boss_health = {boss: 0x1C if boss != 12 else 0x1C * 2 for boss in [*range(8), 12]}
+
+        weapon_energy = {key: float(0x1C) for key in weapon_costs}
+        weapon_boss = {boss: {weapon: world.weapon_damage[weapon][boss] for weapon in world.weapon_damage}
+                       for boss in [*range(8), 12]}
+        flexibility = [(sum(1 if weapon_boss[boss][weapon] > 0 else 0 for weapon in range(9)) *
+                        sum(weapon_boss[boss].values()), boss) for boss in weapon_boss if boss != 12]
```

Comment:
Hmm, I don't mean to be super nitpicky I promise, but this code is so hard to understand as an outsider and I think stuff like this would help

You use `weapon_boss[boss]` and `boss` while iterating `for boss in weapon_boss`
Why not

```
flexibility = [
    (sum(weapon_damages[weapon] > 0 for weapon in range(9)) * sum(weapon_damages.values(), boss)
    for boss, weapon_damages in boss_weapon.items()
]
```

I feel like the use of `for key, value in dict1.items()` could help the readability of these comprehensions a lot here, it makes the code shorter **and** it allows you to name the value

*low priority, I guess*